compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -698,7 +699,7 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MirNeighborCollector<'a, 'tcx> {
                 // This could also be a different Unsize instruction, like
                 // from a fixed sized array to a slice. But we are only
                 // interested in things that produce a vtable.
-                if target_ty.is_trait() && !source_ty.is_trait() {
+                if (target_ty.is_trait() || target_ty.is_dyn_star()) && !source_ty.is_trait() {
```

Comment:
Can you say a little more about why you think this? I'm not necessarily arguing, and I it definitely feels more correct to pair `is_trait` with `is_trait` and `is_dyn_star` with `is_dyn_star`, but I'd like to be able to explain to myself more concretely why one is right.

Reply:
Actually, now that I think about it, this might not be right either. We're basically trying to determine whether or not we need to generate a vtable for a type. We shouldn't ever be emitting a DynStar cast if the target isn't `is_dyn_star`, so the `target_ty.is_trait` is not needed I think? And we don't need to visit to generate a vtable if the source is `dyn*`. So I think maybe what we want here is `!source_ty.is_dyn_star() && target_ty.is_dyn_star()`?

Maybe think about it a bit more, would be happy if you proved me wrong hehe

Reply:
Yeah, that's a good point. This is probably code I copied from somewhere else. I think we can `assert!(target_ty.is_dyn_star())` for sure. I think at least for now we should disallow `source_ty.is_dyn_star()`, but it seems like that will still need a vtable.

To make things more explicit, I think `source_ty.is_dyn_star()` comes up in a case like:
```
trait Foo {
    fn foo(&self);
}

trait Bar: Foo {
    fn bar(&self);
}

let x = make_bar() as dyn* Bar;
let y = x as dyn* Foo;
```
For right now at least, I don't think I know enough about trait inheritance and vtables to make a good call on the right thing to do here, so I propose adding a fixme.

Reply:
Oh, also the comment about Unsizing doesn't make sense because dyn* is sized and we can only cast sized values into it, so there's no unsizing anywhere around here.